### PR TITLE
feat(Folder) #MAG-286 : handle share folders in back 

### PR DESCRIPTION
**Pull request by Rseuret**
## Describe your changes

Handle share for folders 

## Checklist tests
Check if a right is apply to all the folder s children
Check if the old rights are still saved with the new rights applied 

## Issue ticket number and link
https://jira.support-ent.fr/browse/MAG-286

## Checklist before requesting a review (magic string, indentation, comment/documentation...)

##API descriptions : 

https://confluence.support-ent.fr/display/RN/Share

- [x] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

**Review by alicedraillard — APPROVED** [review]

File: src/main/java/fr/cgi/magneto/controller/ShareController.java [hunk 1]
```diff
@@ -44,15 +58,15 @@ public void initManagerRight(final HttpServerRequest request) {
...
+    @Get("/:type/share/json/:id")
     @ApiDoc("Share board by id")
     @ResourceFilter(GetSharesRight.class)
     @SecuredAction(value = "", type = ActionType.RESOURCE)
     public void shareBoard(final HttpServerRequest request) {
         shareJson(request, false);
     }
 
-    @Put("/share/resource/:id")
+    @Put("/:type/share/resource/:id")
```
Comment (flmartineau) [inline src/main/java/fr/cgi/magneto/controller/ShareController.java:69]: Pense à mettre à jour les docs d'API

File: src/main/java/fr/cgi/magneto/service/FolderService.java [hunk 2]
```diff
@@ -67,4 +68,23 @@ public interface FolderService {
...
+         Future<List<String>> getChildrenBoardsIds(String id);
+
+        /**
+         * Share folders
+         * @param id
+         * @param newShares
+         * @param deletedShares
+         * @return
+         */
+        Future<Void> shareFolder(String id, List<SharedElem> newShares, List<SharedElem> deletedShares);
```
Comment (flmartineau) [inline src/main/java/fr/cgi/magneto/service/FolderService.java:89]: Si possible, tu peux ajouter une description pour les paramètres.

File: src/main/java/fr/cgi/magneto/service/impl/DefaultFolderService.java [hunk 3]
```diff
@@ -371,5 +373,63 @@ public Future<JsonObject> updateOldFolder(List<String> boardIds) {
...
+    public Future<List<String>> getChildrenBoardsIds(String id) {
+        Promise<List<String>> promise = Promise.promise();
+
+        List<String> ids = new ArrayList<>();
+        ids.add(id);
+        getFolderChildrenIds(ids)
+                .compose(this::getBoardIdsInFolders)
+                .onSuccess(promise::complete)
+                .onFailure(error -> {
+                    String message = String.format("[Magneto@%s::getAllBoardsIds] Failed to recovers boards and folders ids",
```
Comment (flmartineau) [inline src/main/java/fr/cgi/magneto/service/impl/DefaultFolderService.java:401]: getChildrenBoardsIds dans le log

File: src/main/java/fr/cgi/magneto/service/impl/DefaultShareService.java [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+
+    @Override
+    public Future<JsonObject> upsertSharedArray(String id, JsonObject share, String collection, boolean checkOldRights) {
+        Promise<JsonObject> promise = Promise.promise();
+        JsonArray newShare = toMongoBasicShareFormat(share);
+        JsonObject query = new JsonObject().put(Field._ID, id);
+
+        if (checkOldRights) {
+            getOldDataToUpdate(id, collection).onSuccess(success -> {
+                List<SharedElem> sharedElems = getOldRights( success.getJsonArray("shared", new JsonArray()), newShare);
```
Comment (flmartineau) [inline src/main/java/fr/cgi/magneto/service/impl/DefaultShareService.java:37]: Field.SHARED

File: src/main/java/fr/cgi/magneto/service/impl/DefaultShareService.java [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+        query.match(new JsonObject().put(Field._ID, folderId));
+        mongoDb.command(query.getAggregate().toString(), MongoDbResult.validResultHandler(resultMongo -> {
+            if (resultMongo.isRight()) {
+                JsonObject result = resultMongo.right().getValue()
+                        .getJsonObject(Mongo.CURSOR, new JsonObject())
+                        .getJsonArray(Mongo.FIRSTBATCH, new JsonArray())
+                        .getJsonObject(0);
+                promise.complete(result);
+            } else {
+                String message = String.format("[Magneto@%s::updateOldFolder] Failed to update old folder",
```
Comment (flmartineau) [inline src/main/java/fr/cgi/magneto/service/impl/DefaultShareService.java:172]: getOldDataToUpdate dans le log

File: src/main/java/fr/cgi/magneto/controller/ShareController.java [hunk 6]
```diff
@@ -61,36 +75,93 @@ public void shareResource(final HttpServerRequest request) {
...
+                    } else if (type.equals(Field.FOLDER)) {
+                        handleShareFolder(request, user, newSharedElem, id, i18nHelper, share);
+                    }
                 });
             }
         });
     }
 
+    private void handleShareBoard(HttpServerRequest request, UserInfos user, JsonObject share, String id, List<SharedElem> newSharedElem, I18nHelper i18nHelper) {
+        this.folderService.getFolderByBoardId(id).onSuccess(s -> {
```
Comment (FlorentMr) [inline src/main/java/fr/cgi/magneto/controller/ShareController.java:115]: Manque un onFailure

File: src/main/java/fr/cgi/magneto/controller/ShareController.java [hunk 7]
```diff
@@ -61,36 +75,93 @@ public void shareResource(final HttpServerRequest request) {
...
+                unauthorized(request, "Can't apply this rights");
+            }
+        });
+    }
+
+    private void handleShareFolder(HttpServerRequest request, UserInfos user, List<SharedElem> newSharedElem, String id, I18nHelper i18nHelper, JsonObject share) {
+
+        Future<List<SharedElem>> deletedRightFuture = this.magnetoShareService.getDeletedRights(id, newSharedElem, CollectionsConstant.FOLDER_COLLECTION);
+        this.magnetoShareService.checkParentRights(id, newSharedElem, CollectionsConstant.FOLDER_COLLECTION)
+                .onSuccess(checkRight -> {
```
Comment (FlorentMr) [inline src/main/java/fr/cgi/magneto/controller/ShareController.java:153]: Manque un onFailure

File: src/main/java/fr/cgi/magneto/service/impl/DefaultCardService.java [hunk 8]
```diff
@@ -142,7 +142,6 @@ private Future<JsonObject> syncDocumentRights(String boardId, String documentId,
                         String boardOwnerId = board.getOwnerId();
                         JsonObject shared = WorkspaceHelper.toAPIShareFormat(board.getShared(), cardOwnerId.equals(boardOwnerId), boardOwnerId);
 //                        JsonArray workspaceShare = WorkspaceHelper.toMongoWorkspaceShareFormat(shared);
-                        //TODO BIEN CHECK SI CA MARCHE une fois MAG-286 implentée
```
Comment (FlorentMr) [inline src/main/java/fr/cgi/magneto/service/impl/DefaultCardService.java:145]: Alors ça marche? (faut le laisser pour plus tard ou commentaire à enlever?)

File: src/main/java/fr/cgi/magneto/service/impl/DefaultShareService.java [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+                promise.fail(results.left().getValue());
+            }
+            promise.complete(results.right().getValue());
+        }));
+    }
+
+
+    @Override
+    public  List<SharedElem> getOldRights(JsonArray oldShared, JsonArray newShare) {
+        List<SharedElem>  oldSharedElems= getSharedElemList(oldShared);
```
Comment (FlorentMr) [inline src/main/java/fr/cgi/magneto/service/impl/DefaultShareService.java:94]: ça me trigger trop dsl mais ya un espace en trop

File: src/main/java/fr/cgi/magneto/service/impl/DefaultShareService.java [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+        List<SharedElem> newSharedElems = getSharedElemList(newShare);
+
+        return getOldRights(oldSharedElems, newSharedElems);
+    }
+
+    @Override
+    public  List<SharedElem> getOldRights(JsonArray oldShared, List<SharedElem> newShare) {
+        List<SharedElem>  oldSharedElems= getSharedElemList(oldShared);
```
Comment (FlorentMr) [inline src/main/java/fr/cgi/magneto/service/impl/DefaultShareService.java:102]: là en a un en trop et un oublié pour le =

File: src/main/java/fr/cgi/magneto/service/impl/DefaultShareService.java [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+
+        return getOldRights(oldSharedElems, newShare);
+    }
+    @Override
+    public  List<SharedElem> getOldRights(List<SharedElem> oldSharedElems, List<SharedElem> newSharedElems) {
```
Comment (FlorentMr) [inline src/main/java/fr/cgi/magneto/service/impl/DefaultShareService.java:107]: espace en troooooop

File: src/main/java/fr/cgi/magneto/service/impl/DefaultShareService.java [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+        getOldDataToUpdate(id, boardCollection).onSuccess(success ->{
+            promise.complete(getOldRights(success.getJsonArray(Field.SHARED),newSharedElem));
+        });
+        return promise.future();
+    }
+
+    @Override
+    public Future<Boolean> checkParentRights(String id, List<SharedElem> newSharedElem, String collection) {
+        Promise<Boolean> promise = Promise.promise();
+        this.getOldDataToUpdate(id, collection).onSuccess(folder -> {
```
Comment (FlorentMr) [inline src/main/java/fr/cgi/magneto/service/impl/DefaultShareService.java:193]: manque un onFailure

File: src/main/java/fr/cgi/magneto/controller/ShareController.java [hunk 13]
```diff
@@ -61,36 +75,93 @@ public void shareResource(final HttpServerRequest request) {
...
+                                    JsonObject pushNotif = new JsonObject()
+                                            .put(Field.TITLE, "push.notif.magneto.share")
+                                            .put(Field.BODY, user.getUsername() + " " + i18nHelper.translate("magneto.shared.push.notif.body"));
+                                    params.put(Field.PUSHNOTIF, pushNotif);
+
+                                    List<String> ids = new ArrayList<>();
+                                    ids.add(id);
+                                    this.boardService.shareBoard(ids, share, false).onSuccess(success -> {
+                                        notification.notifyTimeline(request, "magneto.share_folder", user, new ArrayList<>(), id, "test",
+                                                params, true);
```
Comment (flmartineau) [inline src/main/java/fr/cgi/magneto/controller/ShareController.java:139]: C'est quoi ce "test" ?

File: src/main/java/fr/cgi/magneto/controller/ShareController.java [hunk 14]
```diff
@@ -61,36 +75,93 @@ public void shareResource(final HttpServerRequest request) {
...
+                                            params.put(Field.USERNAME, user.getUsername());
+                                            params.put(Field.BOARDURL, "/magneto#/");
+
+                                            JsonObject pushNotif = new JsonObject()
+                                                    .put(Field.TITLE, "push.notif.magneto.share")
+                                                    .put(Field.BODY, user.getUsername() + " " + i18nHelper.translate("magneto.shared.push.notif.body"));
+                                            params.put(Field.PUSHNOTIF, pushNotif);
+
+                                            notification.notifyTimeline(request, "magneto.share_folder", user, new ArrayList<>(), id, "test",
+                                                    params, true);
```
Comment (flmartineau) [inline src/main/java/fr/cgi/magneto/controller/ShareController.java:172]: pareil ici ?

**Review by flmartineau — APPROVED** [review]

File: src/main/java/fr/cgi/magneto/service/ServiceFactory.java [hunk 15]
```diff
@@ -76,6 +82,9 @@ public CommentService commentService() {
     public BoardAccessService boardViewService(){
         return new DefaultBoardAccessService(CollectionsConstant.BOARD_VIEW_COLLECTION,mongoDb);
     }
+    public ShareService shareService(){
+        return new DefaultShareService( mongoDb);
```
Comment (FlorentMr) [inline src/main/java/fr/cgi/magneto/service/ServiceFactory.java:86]: espace

**Review by FlorentMr — CHANGES_REQUESTED** [review]

**Review by FlorentMr — APPROVED** [review]

File: src/main/java/fr/cgi/magneto/Magneto.java [hunk 16]
```diff
@@ -26,7 +26,7 @@ public void start() throws Exception {
 
 		Storage storage = new StorageFactory(vertx, config).getStorage();
 		MagnetoConfig magnetoConfig = new MagnetoConfig(config);
-		ServiceFactory serviceFactory = new ServiceFactory(vertx, storage, magnetoConfig, Neo4j.getInstance(), Sql.getInstance(), MongoDb.getInstance());
+		ServiceFactory serviceFactory = new ServiceFactory(vertx, storage, magnetoConfig, Neo4j.getInstance(), Sql.getInstance(), MongoDb.getInstance() ,securedActions);
```
Comment (alicedraillard) [inline src/main/java/fr/cgi/magneto/Magneto.java:29]: espace apres la virgule pas avant

File: src/main/java/fr/cgi/magneto/service/impl/DefaultFolderService.java [hunk 17]
```diff
@@ -371,5 +373,68 @@ public Future<JsonObject> updateOldFolder(List<String> boardIds) {
         return promise.future();
     }
 
+    @Override
+    public  Future<JsonObject> getFolderByBoardId(String boardId){
```
Comment (alicedraillard) [inline src/main/java/fr/cgi/magneto/service/impl/DefaultFolderService.java:377]: espace avant le {

**Review by alicedraillard — APPROVED** [review]

**Pull request merged**